Source/Dafny/Compiler.cs

```diff
@@ -3122,6 +3131,11 @@ public class CheckHasNoAssumes_Visitor : BottomUpVisitor
       return wrOuter;
     }
 
+    bool IsTailRecursiveByMethodCall(FunctionCallExpr fce) {
```

Comment:
The reason is that the ordinary target of a `FunctionCallExpr` is a function, whereas the enclosing context for a function-by-method is a method. So, one has to look at `fce.Function.ByMethodDecl` to figure out if it's tail recursive. Moreover, this has to be done one level up in the case of `return` statements, or otherwise the compiler emits unreachable code that the Java compiler will flag as an error.